testsuite/demo/run-pass/channel.verona

```diff
@@ -0,0 +1,156 @@
+/*****************************************************************************
+ * This example demonstrates how one could build a channel.
+ *
+ * It illustrates creating and using promises and cowns and communicating
+ * between behaviours.
+ *****************************************************************************/
+
+use "library/queue.verona"
+
+/*
+ * Create a channel from two queues:
+ *   - A queue of reads
+ *   - A queue of writes
+ *
+ * Writing a value to the channel services the first waiting read
+ * if one doesn't exist then the write is queued.
+ *
+ * Reading a value from the channel returns a promise of a value.
+ * If a write has been queued then the read is immediately serviced,
+ * otherwise the read is queued.
+ */
+class Channel[T] {
+  _ws: Queue[Writer[T]] & iso;
+  _rs: Queue[Reader[T]] & iso;
+
+  create(): cown[Channel[T]] & imm {
+    var result = new Channel;
+    result._ws = Queue.create(0);
+    result._rs = Queue.create(1);
+    cown.create(result)
+  }
+
+  /*
+   * Write is a synchronous method.
+   * The method schedules a behaviour to log a write
+   * to the channel.
+   */
+  write(channel: cown[Channel[T]] & imm, value: T) {
+    var cell = Cell.create(value);
+    when(channel) {
+      match((mut-view (channel._rs)).remove()) {
+        var _: None => (mut-view (channel._ws)).add(Writer.create(cell)),
+        var r: Reader[T] => r.write(cell),
+      }
+    }
+  }
+
+  /*
+   * Read is an asynchronous method.
+   * The method returns the wait part of a promise
+   * that will be fulfilled when a value is available.
+   */
+  read(channel: cown[Channel[T]] & imm): cown[Cell[T]] & imm {
+    var r = Reader.create();
+    var handle = (mut-view r).await();
```

Comment:
This is nice as the code can return before this has even entered the promise in the queue, explain this.